### PR TITLE
[WIP] Python APIs for OPTIMIZE ZORDER BY

### PR DESCRIPTION
**Pull request by ahlag**
## Description

Resolves https://github.com/delta-io/delta/issues/1185
Added Python API for executeZOrderBy in `DeltaOptimizeBuilder`

## How was this patch tested?

Added `test_optimize_zorder` to test_deltatable.py

## Does this PR introduce _any_ user-facing changes?

New API for Python executeZOrderBy

### CODE REVIEW

File: python/delta/tests/test_deltatable.py [hunk 1]
```diff
@@ -912,6 +912,26 @@ def optimize() -> None:
...
+            # create DeltaTable
+            dt = DeltaTable.forPath(self.spark, self.tempFile)
+
+            # execute Z-Order Optimization
+            optimizer = dt.optimize()
+            res = optimizer.executeZOrderBy(["a", "b"])
+
+            # assertions
+            self.assertTrue(isinstance(optimizer, DeltaOptimizeBuilder))
+            self.assertTrue(isinstance(res, DataFrame))
```
Comment (vkorukanti) [inline python/delta/tests/test_deltatable.py:933]: This is a good starting point for tests. You can expand this test to verify that the returned DataFrame contains the expected values. Given there are 3 files in the input, we expect a 1 one file as output (as the total size is less than the maxFileSize) (Example verification for optimize compaction is [here](https://github.com/delta-io/delta/blob/master/python/delta/tests/test_deltatable.py#L853)). And we need to verify that there are zOrder metrics (Scala based tests for verification are [here](https://github.com/delta-io/delta/blob/master/core/src/test/scala/org/apache/spark/sql/delta/optimize/OptimizeMetricsSuite.scala#L236)). You could actually copy [this](https://github.com/delta-io/delta/blob/master/core/src/test/scala/org/apache/spark/sql/delta/optimize/OptimizeMetricsSuite.scala#L236) test.

Additional tests needed:
- test that does ZOrder By and also contains WHERE clause.
- a negative test. You could write a test where we pass a partition column as z-order by columns. Scala version of the test is [here](https://github.com/delta-io/delta/blob/master/core/src/test/scala/org/apache/spark/sql/delta/optimize/OptimizeZOrderSuite.scala#L58).

**Review by vkorukanti — COMMENTED** [review]
Thanks @ahlag for the PR. Looks good. Provided comments on the tests.

**Comment by vkorukanti** [comment]
Thank you @ahlag for working on this. Sorry about the python setup issues. Updated the code to fix the errors and improve tests. PR is here: https://github.com/delta-io/delta/pull/1226/files. The first commit contains all the code from this PR.

**Comment by ahlag** [comment]
@vkorukanti 
Thank you so much! Can I ask why was the test failing?

**Comment by vkorukanti** [comment]
> @vkorukanti Thank you so much! Can I ask why was the test failing?

Has lint errors and the expected result need to be updated. See the changes [here](https://github.com/delta-io/delta/pull/1226/commits/a19ccc9e3750285a4bf70b87a624c19883ce2ec2).

**Comment by ahlag** [comment]
@vkorukanti 
Thank you so much!

**Comment by vkorukanti** [comment]
@ahlag Closing this PR. The other PR is merge [here](https://github.com/delta-io/delta/commit/6328adf09e1c9645230ce9abd8eacac292887342). Thank you for the contribution.

**Pull request closed without merging**